### ISSUE TITLE
XDG Base Directory Specs

### ISSUE THREAD

**Issue by evgenyz** [minetest/minetest#864]
Please consider supporting freedesktop.org's specs about config directory location. References: http://freedesktop.org/wiki/Software/xdg-user-dirs/, http://standards.freedesktop.org/basedir-spec/basedir-spec-latest.html#variables.

**Comment by Wuzzy2** [comment]
Is it possible to add this to 0.5.0 milestone or is it already too late for this?

**Comment by nerzhul** [comment]
it's not late we are not in feature freeze, the problem is to migrate data

**Comment by Ferk** [comment]
It wouldn't be impossible to keep some backwards compatibility.
If the folder in ~/.minetest exists, use it, otherwise use the new locations.

Git for example does that too, it can read from both ~/.config/git/config and ~/gitconfig

**Comment by paramat** [comment]
c55 disapproves of XDG but wrote:

> However, I support moving the media cache to ~/.cache/minetest/, because that way backup software, users that have too little disk space and other things have easier time handling it in an appropriate way.

Anyone agree with that?
Generally i disapprove of changing everything according to some external standard, if we change anything let's just do a minimal amount of things that are genuinely helpful.
Remember that people are used to how things are, whether how it is makes sense according to some external standard is irrelevant.
So :-1: for the original request but i'm ok with anything that is very helpful and not disruptive.

**Comment by Wuzzy2** [comment]
A key reason why XDG exists is to stop programs polluting the home directory with lots of directories starting with a period. Supporting XDG is a good thing, not just for the sake of supporting XDG.
Many other applications follow XDG; XDG is clearly not just any random irrelevant standard you can easyly dismiss. Rejecting XDG only because “we're different” seems to be a poor counter-argument.

Also, you can't use the “disruptive” argument right now because 0.5.0 *can* be a breaking release. :P
Also, Minetest is technically still in alpha. Being so conservative in the alpha stage of development is brain dead.

Now is the perfect time to do stuff like this, especially because of 0.5.0. If we reach 1.0.0, and THEN try to switch to XDG, it will be much more difficult.

**Comment by nerzhul** [comment]
@Wuzzy2 breaking homedir is not possible, but migrate user data is totally possible

**Comment by rubenwardy** [comment]
> Also, Minetest is technically still in alpha. Being so conservative in the alpha stage of development is brain dead.
> Now is the perfect time to do stuff like this, especially because of 0.5.0. If we reach 1.0.0, and THEN try to switch to XDG, it will be much more difficult.

Minetest doesn't use semvar. 0.5.0 is basically just 5.0.0. And 0.4.16 is 4.16.0

**Comment by Wuzzy2** [comment]
> Minetest doesn't use semvar. 0.5.0 is basically just 5.0.0. And 0.4.16 is 4.16.0

And what will 1.0.0 be? Infinity.0.0?
Perfect logic! xD

Anyways, I don't care. I consider Minetest to be alpha stage software as there are many features missing, Lua API and Lua API documentation are incomplete and there are many important bugs to fix. I thought that was very obvious.

Using version 1.0 (or similar) as a major milestone is actually very common practice virtually everywhere and has nothing to do with SemVer. Many people associate versions starting with 0 as imcomplete/alpha/beta/buggy. Definitely true for Minetest now but I'm sure Minetest will become very mature and stable (“1.0.0-worthy”) in the future. :-)

**Comment by paramat** [comment]
Wuzzy2 as i wrote:

> i'm ok with anything that is very helpful

and "stop programs polluting the home directory with lots of directories starting with a period." is a good idea. I'm just saying blindly doing things we don't need to for the sake of following a standard is not good.

> you can't use the “disruptive” argument right now because 0.5.0 can be a breaking release

Yes i can , because 0.5.0 does not mean we can change and rewrite everything.

> Minetest is technically still in alpha
If we reach 1.0.0,

'Alpha' and '1.0.0' is meaningless in MT, it has been usable for years and has open-ended dev, so is not really 'alpha' or 'beta'. This '1.0.0' idea is meaningless too for the same reasons. If 1.0.0 is used at some point it will mean nothing, it's impossible to pick a point where it is 'finished', there will always be more to do. MT has been 'complete' for years but also will always be incomplete. It's not a commercial release so 'alpha' 'beta' '1.0.0' don't mean much.

> Lua API and Lua API documentation are incomplete and there are many important bugs to fix

The API will always be incomplete, there will always bo more to add. There will always be bugs.

**Comment by paramat** [comment]
@numberZero maybe you would be interested in doing what celeron55 suggests?:

> However, I support moving the media cache to ~/.cache/minetest/, because that way backup software, users that have too little disk space and other things have easier time handling it in an appropriate way.

**Comment by paramat** [comment]
:-1: for XDG.
c55 disapproves of XDG but wrote:

> However, I support moving the media cache to ~/.cache/minetest/, because that way backup software, users that have too little disk space and other things have easier time handling it in an appropriate way.

Any core dev support doing this?

**Comment by jastevenson303** [comment]
Was done a long time ago.

**Comment by paramat** [comment]
Ok good. Issue is 5 years old with 3 disapprovals.

**Comment by eli-schwartz** [comment]
Repeating my comment from https://github.com/minetest/minetest/issues/8151#issuecomment-460102567 (which is closed as a duplicate):

Failure to follow the XDG Base Directory specification **or its OS-specific adaptations** results in software which is just as broken on macOS **and** Windows as it is on Linux.

On macOS, `XDG_CONFIG_HOME` maps to `~/Library/Preferences/`, `XDG_DATA_HOME` maps to `~/Library/Application Support/`, and `XDG_CACHE_HOME` maps to `~/Library/Caches/`.

On Windows, `XDG_CONFIG_HOME` and `XDG_DATA_HOME` are not separated, as both are in `%APPDATA%`, but `XDG_CACHE_HOME` must be in `%LOCALAPPDATA%` to prevent roaming across network shares (it does not specify when you can clean it up though).

Linux of course provides separate directories for all three (just like macOS), and goes even further by allowing them to be relocated via environment variables. But I hardly see why the lack of cross-platform support for overriding `XDG_*` should cast doubt on the usefulness of a program intelligently deciding where to put files according to the well-documented, preferred OS-specific locations.

**Comment by benrob0329** [comment]
@paramat I hardly think that a statement made by a long-since inactive dev 5 years ago qualifies as a disapproval at this point.

I count two devs against (paramat and rubenwardy), two known to be for (nerzhul and Calinou), and quite a few users for.

C55 was against it because it supposedly made it more difficult to jump into the conf dir, however I think that with the brand new "bookmarks" and/or "places" feature added in KDE 1.0 that is negligible.

What is not negligible is having a hundred dot files in my home dir, and software (like Minetest!) refusing to follow standard data locations continuing to clog it up.

This isn't 2012, storing data directly in $HOME/.dir simply isn't acceptable anymore. The XDG standard was made for a reason, and its about time we listened to it.

**Comment by nikp123** [comment]
Also, you don't have to make following XDG mandatory (c55s argument - where everyone is forced to follow it, thus complicating everything), as you can simply after checking that the user files are in ``$HOME/.minetest`` and if they're not, simply check ``$XDG_DATA_HOME/minetest`` (by default ~/.local/share) afterwards (Unity does this for every game, as far as I know).

Also for simplicity sakes, you don't even have to tear up the directory, just keep everything in ~/.local/share/minetest as many applications do.

**Comment by paramat** [comment]
benrob0329,
Just saw your comment.
>  I hardly think that a statement made by a long-since inactive dev 5 years ago qualifies as a disapproval at this point. I count two devs against (paramat and rubenwardy), two known to be for (nerzhul and Calinou), and quite a few users for.

It does not matter how long inactive kahrl is, keep in mind that Celeron55 is long inactive too.
Celeron55 also commented at the same time kahrl did, so you are implying we should dismiss their opinion too.
Kahrl has the highest number of contributions of the very earliest core devs, so is a very significant core dev.
This issue is also very old and kahrl's disapproval was given on the issue after the issue was opened, so it is not irrelevant to the issue.
Rubenwardy has not disapproved in this thread. Nerzhul has not stated support in this thread. Calinou is not an engine core dev. So you are very incorrect.
Users do not make development decisions or decide which requests are closed.
So even if we ignore kahrl, this has 2 core dev disapprovals and no core dev support after 7 years, so the end result is no different.

**Comment by mazunki** [comment]
This is closed, but I am going to leave a comment anyway. I am new to Minetest, but I wouldn't mind actually "fixing" the code to follow XDG Base Directory standards.

I don't see a single reason why one would pollute the home directly when there is a well-known and well-used standard out there. 

A [blogpost](https://0x46.net/thoughts/2019/02/01/dotfile-madness/) explains the issue very well. Making a statement against this, without reason, seems like a reason for me to search for something else instead. 
I agree with @celeron55 's suggestion for the folders: 

> ~/.minetest/cache -> ~/.cache/minetest/
> ~/.minetest/client/serverlist -> ~/.local/share/minetest/client/serverlist
> ~/.minetest/debug.txt -> ~/.local/share/minetest/debug.txt (?)
> ~/.minetest/minetest.conf -> ~/.config/minetest/minetest.conf
> ~/.minetest/worlds -> ~/.local/share/minetest/worlds
> ~/.minetest/games -> ~/.local/share/minetest/games
> ~/.minetest/mods -> ~/.local/share/minetest/mods

except I would maybe put debug.txt into cache, and serverlist into config (although this is debatable). 

## tl;dr: 
Why reinvent the wheel? Why adhere to *no* standard when one could adhere to _one_ well-made one?

**Comment by okias** [comment]
... 3 years; The XDG spec should be followed on Linux, it makes life much easier to use the right directories on Linux.

Usually applications have a transition in-place to fluently switch.

I would suggest
 1. enable paths for setups where no `~/.minetest` exists and keep old in `~/.minetest`
 2. (possible later) migrate paths from `~/.minetest` to new locations (can replace point 1. or happen with later releases)
 3. (optionally) symlink `~/.minetest/xy` to new locations (for transition period (?) )
